### ISSUE TITLE
no way to define "no option forwardfor"

### ISSUE THREAD

**Issue by hoerup** [haproxy/haproxy#130]
currently "option forwardfor" cant't be unset, so there is no way to enable it in the default section and disable it in a particular frontend

**Comment by lukastribus** [comment]
That's not a bug, but documented:

> Keywords marked with "(*)" can be optionally inverted using the "no" prefix, e.g. "no option contstats".

`option forwardfor` does not have a `(*)`.

Transforming this into a feature request.

**Comment by onetwopunch** [comment]
Any update on this? It seems odd that you can set other http settings as default, then override, but not this one. This requires adding a `forwardfor` option to several http front/backends just to allow for one new tcp mode block.

**Comment by lukastribus** [comment]
> This requires adding a `forwardfor` option to several http front/backends just to allow for one new tcp mode block.

No, you can work with multiple default sections, that is way easier and you can even setup logging in a sane way

https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#4

```
defaults
mode http
option httplog
option forwardfor

# put all http sections here

defaults
mode tcp
option tcplog


# put all tcp sections here

```

**Comment by hervenicol** [comment]
> currently "option forwardfor" cant't be unset, so there is no way to enable it in the default section and disable it in a particular frontend

It seems adding `option forwardfor except 0.0.0.0/0` in a specific section could do the job.

**Comment by VigneshSP94** [comment]
@lukastribus @TimWolla can I work on this ticket ? I have never written code in C, but would like to get started with contributing to haproxy. If this still needs fix and I can get some help, I will happily work on it.

**Comment by wtarreau** [comment]
@VigneshSP94 sure, feel free to have a look at it! Please take the time to read CONTRIBUTING and doc/coding-style.txt to avoid initial mistakes, this should be of great help. The parts that you'll be interested in are located in src/cfgparse-listen.c.

**Comment by hcgpalm** [comment]
Apart from beeing quite confusing, specifying multiple default sections doesn't work if you're using config directories to separate defaults from specifics into separate files.

Perhaps the question that needs to be asked here is - What is the use case? Why would you want to disable option forwardfor for a particular section?

Although I can certainly imagine there are situations when you'd actually want to disable it for a particular mode http section, for me, it's about getting rid of the annoying warning at startup:
"config : 'option forwardfor' ignored for frontend 'x' as it requires HTTP mode"

Maybe removing that warning would take us a long way?
There are numerous other options that only apply under certain conditions that haproxy doesn't warn about, so what's so special about "option forwardfor"?

**Comment by ronaldocheruiyot** [comment]
Hi @wtarreau  @lukastribus ,

I'm working on a patch to resolve this issue and implement support for `no option forwardfor`.

My plan is to update the keyword parser in `src/cfgparse-listen.c`. When the parser catches the inverted no variant (checking for a `kwm == KWM_NO` modifier or similar context flag), I will clear the `PR_O_FWDFOR` bitmask flag from `curproxy->options` to effectively unset the inherited default configuration. I will also make sure to clean up or safely bypass any extra arguments like `except` or `header` if they accompany a `no` directive.

I'll ensure it respects the guidelines in `doc/coding-style.txt` and submit a clean PR shortly. Let me know if you have any initial feedback on this direction!
